dieboljo/embedded-loop-station
# Feature/basic loop

## Related Issues:
* [Record Monophonic Audio](https://app.asana.com/0/1203221076104575/1203350230081701)
* [Implement Loop Functionality](https://app.asana.com/0/1203221076104575/1203349316387105)
* [Replace Mode](https://app.asana.com/0/1203221076104575/1203718223981527)

## Main Changes:
* Incorporated the **Track** class into the main source code.
* Fixed some offset issues
* Fixed some _audio interrupt_ issues, where state changes weren't occurring synchronously. Might be a bit overprotective in some cases, but we can dial back as needed in the future.

## Steps To Test:
1. Checkout this branch, then compile and upload the program to your Teensy device.
2. Press the record button, and record a sample.
3. Press play, and confirm that the audio plays in a loop.
4. During playback, press record again to "punch in" and replace existing audio at that point. Punch back out by pressing play or stop. Pressing record again will pause the recording and playback.
5. Run the unit tests written for the **Track** class.

## NOTES
- After looking at the functionality of "professional" loop pedals, I think the behavior of this should be changed. Mainly, for each track, the initial recording should set the length of the loop. That way, additional replacements won't extend the length, which causes odd behavior.
  - To do this, we will need to alter the button behavior: the record button will simply punch in and out of recording (it won't affect play/pause/stop status), the play button will toggle play/pause, and stop will stop playback, return to the start of the track, and undo any replacements or overdubs applied to the current loop iteration (for the track).
- There are some sound artifacts that crop up when replacing audio. I think we will need to switch to the Serial Flash library and rotate three tracks (read, write, and erase), so that each new write is written to a fresh track, with previous audio fed in alongside through a track mixer. This is also necessary for overdub, due to the read/write constraints of SD cards.

Despite these issues, I think this is a good base from which we can continue: the public methods of the Track class should allow for us to start linking up the display with audio functionality.

## TODO:
- [ ] Switch to the **SerialFlash** library
- [ ] Change the behavior of the buttons, as noted above
- [ ] Rotate tracks only at the end of the predetermined loop length (unless the stop button is pressed, in which case the write and erase tracks will be swapped)
